shared-types.d.ts

```diff
@@ -0,0 +1,220 @@
+// TypeScript Version: 3.0
+
+/**
+ * A location in a string or buffer
+ */
+export interface Point {
+  line: number
+  column: number
+  offset: number
+  _index?: number
+  _bufferIndex?: number
+}
+
+/**
+ * A token type
+ *
+ * TODO: enumerate token types
+ */
+export type Type = string
+
+/**
+ *
+ */
+export interface Token {
+  type: Type
+  start: Point
+  end: Point
+
+  previous: Token
+  next: Token
+
+  /**
+   * Declares a token as having content of a certain type.
+   * Because markdown requires to first parse containers, flow, content completely,
+   * and then later go on to phrasing and such, it needs to be declared somewhere on the tokens.
+   */
+  contentType: 'flow' | 'content' | 'string' | 'text'
+
+  // TODO move these to interfaces extending Token, or move these to a `data` property similar to unist (would require code refactor)
+  /**
+   * Used for whitespace in several places which needs to account for tab stops
+   */
+  _size?: number
+
+  /**
+   * ends with a CR, LF, or CRLF.
+   */
+  _break?: boolean
+
+  /**
+   * Used when dealing with linked tokens. A child tokenizer is needed to tokenize them, which is stored on those tokens
+   */
+  _tokenizer?: Tokenizer
+
+  /**
+   * Used for attention (emphasis, strong).
+   *
+   * could be (enter emphasis, enter emphasisMarker, exit emphasisMarker, enter strong, enter strongMarker, exit strongMarker, enter data, exit data)
+   */
+  _events?: Event[]
+
+  /**
+   * This is used for tokens that are already “subtokenized”.
+   *
+   * E.g., when parsing flow, there are content tokens, but those are directly tokenized into definitions/setext/paragraphs
+   */
+  _subevents?: Event[]
+
+  /**
+   * Set to true to mark that a token (e.g., with subevents) is already handled
+   */
+  _contentTokenized?: boolean
+
+  /**
+   * close and open are also used in attention:
+   * depending on the characters before and after sequences (**),
+   * the sequence can open, close, both, or none
+   */
+  _open?: boolean
+
+  /**
+   * close and open are also used in attention:
+   * depending on the characters before and after sequences (**),
+   * the sequence can open, close, both, or none
+   */
+  _close?: boolean
+  _marker?: number
+  _side?: number
+
+  /**
+   * Generally, tabs and spaces behave the same, but in the case of a hard break through trailing spaces ( \n), tabs
+   */
+  _tabs?: boolean
+}
+
+/**
+ *
+ */
+export type Event = [string, Token, unknown]
+
+/**
+ * These these are transitions to update the CommonMark State Machine (CSMS)
+ */
+export interface Effects {
+  /**
+   * Enter and exit define where tokens start and end
+   */
+  enter: (type: Type) => void
+
+  /**
+   * Enter and exit define where tokens start and end
+   */
+  exit: (type: Type) => void
+
+  /**
+   * Consume deals with a character, and moves to the next
+   */
+  consume: (code: number) => void
```

Comment:
Would it be a good idea to make a specific type for the character codes?

It can be any result (positive int) from String.charCodeAt (except NaN). And the following special codes are used:
https://github.com/micromark/micromark/blob/005ad17216ba591560a1a73026c35d2a669fc2a7/lib/character/codes.js#L1-L7

Reply:
It could be a good idea.
Both the known types and character codes could be turned in a union of known types or turned into an [opaque type](https://codemix.com/opaque-types-in-javascript), currently both are transparent primitives.

Reply:
If micromark is written in native TypeScript, we can use TypeScript's `const enum` feature for these constant values. 

By using `const enum`, those import statements while compiling and we don't need to use `script/babel-transform-constants.js` to do that. And we have an easy way to define types for character codes.

However, TypeScript doesn't allow `null` as a value of `const enum`, so we have to do something special for `eof`.